torch/_tensor_docs.py

```diff
@@ -1890,19 +1890,132 @@ def add_docstr_all(method, docstr):
                r"""
 index_put_(indices, values, accumulate=False) -> Tensor
 
-Puts values from the tensor :attr:`values` into the tensor :attr:`self` using
-the indices specified in :attr:`indices` (which is a tuple of Tensors). The
+Puts values from the tensor ``values`` into the tensor :attr:`self` using
+the indices specified in ``indices`` (which is a tuple of Tensors). The
 expression ``tensor.index_put_(indices, values)`` is equivalent to
 ``tensor[indices] = values``. Returns :attr:`self`.
 
-If :attr:`accumulate` is ``True``, the elements in :attr:`values` are added to
+If :attr:`accumulate` is ``True``, the elements in ``values`` are added to
 :attr:`self`. If accumulate is ``False``, the behavior is undefined if indices
 contain duplicate elements.
 
 Args:
     indices (tuple of LongTensor): tensors used to index into `self`.
     values (Tensor): tensor of same dtype as `self`.
     accumulate (bool): whether to accumulate into self
+
+Details:
+
+Each of the tensors in ``indices`` is a set of coordinates into :attr:`self`.
+The first tensor contains coordinates into the first dimension, the second
+contains coordinates into the second dimension, etc. The number of tensors in
+``indices`` is the number of dimensions that each individual coordinate has.
+The tensors must either all have the same shape or their shapes must be
+broadcastable into a common shape (see :ref:`broadcasting-semantics`). One set
+of corresponding elements within the broadcasted index tensors represents one
+index coordinate. For instance::
+
+    indices = [
+        tensor([0, 4, 3, 2]),
+        tensor([5, 6, 8, 9])
+    ]
+
+In this example, ``indices`` contains four 2-D coordinates, (0, 5), (4, 6),
+(3, 8), and (2, 9).
+
+Here is an example with broadcasting::
+
+    indices = [
+        tensor([[2], [5]]),
+        tensor([3, 5, 0])
+    ]
+
+``indices[0].shape`` is (2, 1), and ``indices[1].shape`` is (3,). These shapes
+are broadcastable into the shape (2, 3). When the tensors are expanded into
+this shape, they become::
+
+    indices = [
+        tensor([[2, 2, 2], [5, 5, 5]]),
+        tensor([[3, 5, 0], [3, 5, 0]])
+    ]
+
+Now we can see that ``indices`` in this example represents six 2-D
+coordinates, and they are grouped into two sets of three. The first set is
+(2, 3), (2, 5), (2, 0), and the second set is (5, 3), (5, 5), (5, 0).
+
+If ``self`` has more dimensions than the number of tensors in ``indices``, then
```

Comment:
rather than pretending that more index tensors are appended, isn't it better to say that indices apply to slices of original tensor rather than individual elements? It's a much easier mental (and physical) model for e.g. x[ind] for multidimensional x than expanding indices to cover all dimensions
E.g. numpy has the following:

>It results in the construction of a new array where each value of the index array selects one row from the array being indexed and the resultant array has the resulting shape (number of index elements, size of row).

>In general, the shape of the resultant array will be the concatenation of the shape of the index array (or the shape that all the index arrays were broadcast to) with the shape of any unused dimensions (those not indexed) in the array being indexed.